tools/nemo_forced_aligner/utils/data_prep.py

```diff
@@ -137,157 +244,356 @@ def get_y_and_boundary_info_for_utt(text, model, separator):
         SPACE_ID = model.decoder.vocabulary.index(" ")
 
-        y_token_ids_with_blanks = [BLANK_ID]
-        token_info = [{"text": BLANK_TOKEN, "s_start": 0, "s_end": 0,}]
-        word_info = []
-        segment_info = []
+        utt.token_ids_with_blanks = [BLANK_ID]
+
+        # check for text being 0 length
+        if len(text) == 0:
+            utt.S = len(utt.token_ids_with_blanks)
+            return utt
+
+        # check for # tokens + token repetitions being > T
+        all_tokens = get_char_tokens(text, model)
+        n_token_repetitions = 0
+        for i_tok in range(1, len(all_tokens)):
+            if all_tokens[i_tok] == all_tokens[i_tok - 1]:
+                n_token_repetitions += 1
+
+        if len(all_tokens) + n_token_repetitions > T:
+            utt.S = len(utt.token_ids_with_blanks)
+            logging.info(
+                f"Utterance {utt_id} has too many tokens compared to the audio file duration."
+                " Will not generate output alignment files for this utterance."
+            )
+            return utt
+
+        # build up data structures containing segments/words/tokens
+        utt.segments_and_tokens.append(Token(text=BLANK_TOKEN, text_cased=BLANK_TOKEN, s_start=0, s_end=0,))
 
         segment_s_pointer = 1  # first segment will start at s=1 because s=0 is a blank
         word_s_pointer = 1  # first word will start at s=1 because s=0 is a blank
 
         for i_segment, segment in enumerate(segments):
-            words = segment.split(" ")  # we define words to be space-separated characters
+            # add the segment to segment_info and increment the segment_s_pointer
+            segment_tokens = get_char_tokens(segment, model)
+            utt.segments_and_tokens.append(
+                Segment(
+                    text=segment, s_start=segment_s_pointer, s_end=segment_s_pointer + (len(segment_tokens) - 1) * 2
+                )
+            )
+            segment_s_pointer += len(segment_tokens) * 2 + 2
+
+            words = segment.split(" ")  # we define words to be space-separated substrings
             for i_word, word in enumerate(words):
 
                 # convert string to list of characters
                 word_tokens = list(word)
                 # convert list of characters to list of their ids in the vocabulary
-                word_ids = get_char_tokens(word, model)
-                for token, id_ in zip(word_tokens, word_ids):
-                    # add the text token and the blank that follows it
+                word_token_ids = get_char_tokens(word, model)
+
+                # add the word to word_info and increment the word_s_pointer
+                utt.segments_and_tokens[-1].words_and_tokens.append(
+                    Word(text=word, s_start=word_s_pointer, s_end=word_s_pointer + len(word_tokens) * 2 - 2)
+                )
+                word_s_pointer += len(word_tokens) * 2 + 2  # TODO check this
```

Comment:
There's a lot of todos in this file. It's ok cause we're near freeze but after or is merged all of these need the be updated and or fixed